src/compiler/binder.ts

```diff
@@ -563,20 +611,135 @@ namespace ts {
+            }
+
+            if (file.externalModuleIndicator) {
+                return Diagnostics.Identifier_expected_0_is_a_reserved_word_in_strict_mode_Modules_are_automatically_in_strict_mode;
+            }
+
+            return Diagnostics.Identifier_expected_0_is_a_reserved_word_in_strict_mode;
+        }
+
+        function checkStrictModeBinaryExpression(node: BinaryExpression) {
+            if (inStrictMode && isLeftHandSideExpression(node.left) && isAssignmentOperator(node.operatorToken.kind)) {
+                // ECMA 262 (Annex C) The identifier eval or arguments may not appear as the LeftHandSideExpression of an
+                // Assignment operator(11.13) or of a PostfixExpression(11.3)
+                checkStrictModeEvalOrArguments(node, <Identifier>node.left);
+            }
+        }
+
+        function checkStrictModeCatchClause(node: CatchClause) {
+            // It is a SyntaxError if a TryStatement with a Catch occurs within strict code and the Identifier of the
+            // Catch production is eval or arguments
+            if (inStrictMode && node.variableDeclaration) {
+                checkStrictModeEvalOrArguments(node, node.variableDeclaration.name);
+            }
+        }
+
+        function checkStrictModeDeleteExpression(node: DeleteExpression) {
+            // Grammar checking
+            if (inStrictMode && node.expression.kind === SyntaxKind.Identifier) {
+                // When a delete operator occurs within strict mode code, a SyntaxError is thrown if its
+                // UnaryExpression is a direct reference to a variable, function argument, or function name
+                let span = getErrorSpanForNode(file, node.expression);
+                file.bindDiagnostics.push(createFileDiagnostic(file, span.start, span.length, Diagnostics.delete_cannot_be_called_on_an_identifier_in_strict_mode));
+            }
+        }
+
+        function isEvalOrArgumentsIdentifier(node: Node): boolean {
+            return node.kind === SyntaxKind.Identifier &&
+                ((<Identifier>node).text === "eval" || (<Identifier>node).text === "arguments");
+        }
+
+        function checkStrictModeEvalOrArguments(contextNode: Node, name: Node) {
+            if (name && name.kind === SyntaxKind.Identifier) {
+                let identifier = <Identifier>name;
+                if (isEvalOrArgumentsIdentifier(identifier)) {
+                    // We check first if the name is inside class declaration or class expression; if so give explicit message
+                    // otherwise report generic error message.
+                    let span = getErrorSpanForNode(file, name);
+                    file.bindDiagnostics.push(createFileDiagnostic(file, span.start, span.length,
+                        getStrictModeEvalOrArgumentsMessage(contextNode), identifier.text));
                 }
             }
         }
 
+        function getStrictModeEvalOrArgumentsMessage(node: Node) {

```

Comment:
I started to write it out, and then realized that while it is possible, it seems really silly. It's fine how it is.